### ISSUE TITLE
Custom Headers in Options Rest API Response (Cloud-formation is autogenerated)

### ISSUE THREAD

**Issue by jonmifsud** [aws-amplify/amplify-cli#2308]
**Is your feature request related to a problem? Please describe.**
We have a few instances where we'd like to send some information via headers in conjunction with a get request, where tenant details are passed alongside a request (as a user can have multiple tenant accounts) - If we update the cloud formation template to add this header everything works - however adding/updating the Rest API re-generates the cloud formation template and we lose our changes.

**Describe the solution you'd like**
Update `api-params.json` to allow for additional custom headers to be added. [whilst at it might as well consider doing the same for custom origins]

**Describe alternatives you've considered**
We have been modifying the cloud formation after the generation and re-pushing changes. Then one has to go to the API Gateway and publish to make sure header changes are propagated. 

This works but is prone to mistakes if you forget to repeat the steps.

**Comment by attilah** [comment]
@jonmifsud could you please send details about how you utilize the Amplify CLI, and what steps you've to repeat/do to achieve the correct deployment you're aiming for?

**Comment by jonmifsud** [comment]
Sure @attilah.

So in once instance, we're using the amplify-cli to create a new API Endpoint with a lambda function, which in addition to the user needs to additionally take input to know which tenant the user wants to query. 

This is particularly as we want to keep` get` requests as `get` requests without sending tenant ids within the URL to standardise the API and not expose that in the URL. 

This is also important when we expose some endpoints as public APIs as a single user can manage multiple tenants. The current tenant cannot be stored in a session as there may be requests for different tenants at the same time.

What we end up doing right now is update the cloud formation template to add these within the allowed options response for the endpoint. However every time we add a new path to the API the cloud formation re-builds and we lose the changes having to re-apply them.

Without these headers, chrome would reject the request as the options request does not return this as an accepted header.

I hope this gives a bit more clarity of what we're doing and why.

**Comment by attilah** [comment]
@jonmifsud thanks for the explanation, it is more clear now, I'll discuss this with the team and get back to you.

**Comment by jonmifsud** [comment]
@attilah thank you. Happy to propose PR changes with a bit of direction where things are if it makes sense to make this change.

**Comment by attilah** [comment]
@jonmifsud the more the merrier they say, so your feedback is welcome, go ahead and add it please!

**Comment by kaustavghosh06** [comment]
@jonmifsud Do you need some help with the PR?

**Comment by jonmifsud** [comment]
Yeah actually a couple of pointers would help we still got this issue.

I know we got to modify `apigw-cloudformation-template-default.json.ejs` within `amplify-category-api` package inside the cloud formation templates.

To be able to add a variable value to `method.response.header.Access-Control-Allow-Headers`

What I've done for now is update it to `"method.response.header.Access-Control-Allow-Headers": "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent<%= props.paths[i].headers %>'",`

Which looks like it's working though probably not the cleanest solution.

**Comment by mattfysh** [comment]
Is it just me or does the current definition of `Access-Control-Allow-Headers` prevent you from sending a whole range of headers to the REST API?

**Comment by jonmifsud** [comment]
@kaustavghosh06 / @attilah - not sure if you guys are still on the project. A recent update seems to have broken our work-around completely for this.

Since the `cloudformation-template.json` was moved into the build folder it can no longer be edited to add custom headers... any pointers of how we can get a PR for this submitted?

**Comment by huyttq** [comment]
@jonmifsud use `amplify override api` https://docs.amplify.aws/cli/restapi/override/
